Modernize accounts packages

jamesmillerburgess opened the pull request:
### Update
This pull request modernizes the various `accounts-*` packages.

#### Todo:
- [x] Removal of `underscore` in favor of equivalent ES6+ features
- [x] Check that `_.each` replacements are applied correctly to arrays and objects
- [x] Prefer `includes` over `indexOf`
- [x] Prefer `Array.isArray` over `instanceof Array`
- [x] Removal of `self = this` in all cases where the same scoping can be achieved with arrow functions
- [x] Prefer arrow functions over `function`
- [x] Prefer `class` method declarations over `prototype` manipulation
- [x] Removal of `var` and `function` declarations in favor of `const` and `let`
- [x] Prefer template literals over string concatenation
- [x] Object deconstruction where it makes sense
- [x] Prefer module exports to globals
- [x] Apply DRY for repeated utility functions
- [ ] Extraction of method bodies into logically grouped files
- [ ] JSDoc comments on all methods
- [ ] Additional comments where needed
- [ ] Additional unit tests
- [ ] Additional integration tests
- [x] Increment package numbers

#### Package size chart (as of 1.6.1-beta.3):
```
Package             | Original web.browser | Modernized web.browser | Original web.browser.legacy | Modernized web.browser.legacy
accounts-base       | 12.8  kb             | 11.3  kb               | 12.8  kb                    | 13.3  kb
accounts-password   |  2.16 kb             |  2.32 kb               |  2.16 kb                    |  2.37 kb
```

### Original Comment
This pull request removes the `underscore` dependency from `accounts-base`. It's the first step in a larger modernization/refactoring that I was thinking I could do as a contribution to the Meteor codebase. I made the PR now, because I wanted to get some feedback first on the idea and how far to go.

Options:
1. Remove `underscore` dependencies
2. ES6 syntax/style (e.g. remove usage of `self`, apply Prettier styling, and leverage other ES features)
3. Refactor large functions/files into smaller pieces to enable better test coverage
4. Other...?

Please let me know if this sounds like something that could help and which of the options sound about right.

benjamn commented:
This is awesome! All of the options you mentioned sound good, except that we would rather not reformat code with Prettier. It's a great tool if your whole project is bought into it, but Meteor has its own code styles that are worth preserving (or at least not needlessly changing).

jamesmillerburgess commented:
Great!

I'll switch this pull request to WIP and start a deeper pass over the accounts packages.

jamesmillerburgess commented:
@benjamn 

Since this PR is already going to create a big diff on these files, do you think it is a good time to also update the code style to comply with the AirBnB/Meteor style guide (not Prettier 👍 )?

For example:
- `accounts_client.js` should be named `AccountsClient.js` [22.6](https://github.com/meteor/javascript/#22.6)
- Multiline `if` blocks should either have braces or be reduced to a single line [16.1](https://github.com/meteor/javascript/#16.1)
- Various string stuff [6](https://github.com/meteor/javascript/#strings)

hwillson commented:
@jamesmillerburgess The topic of Meteor's code style has been coming up a lot recently. I'll add this to our next issue triage meeting agenda (tomorrow), and get back to you with an update. My $0.02 though - if we're going to start changing styles to line up with AirBnB/Meteor's style guide, unless we're updating the entire codebase at once, things are likely to get ugly/confusing. If we enforce a new code style for all future changes only, we're going to end up with an even more diverse set of styles, as people make changes only to the new code they're working on, while the old code stays as is. Yes, that's where using something like `prettier` comes in, but incorporating it into the Meteor dev cycle will definitely need to be discussed/planned/tested/etc.

hwillson commented:
@jamesmillerburgess Regarding https://github.com/meteor/meteor/pull/9558#issuecomment-359808310, the plan for now is to keep the code style as is. So please follow the style of the existing source file you're editing (as much as possible), unless of course you see something that's really out of whack (like missing braces, missing spaces, 500 character lines, etc.). That being said, changes to modernize the code (like you're doing) are always welcome - so things like adapting newer ES* syntax, removing underscore, making sure modules are used, etc. The AirBnB/Meteor style guide is not really maintained, and was started based on some original linting rules that were in place with Meteor way back when. It hasn't been updated, and there are many rules in there that aren't used / agreed on. Also, jumping to use something like `prettier` isn't being considered right now.

On a related note, it looks like you've done a tonne of work here - thanks! Do you think things are ready for a review or should we hold off a bit longer?

jamesmillerburgess commented:
@hwillson Thanks for the feedback! I will try to follow those guidelines while editing, and we can later revert anything that seems extraneous.

The PR still isn't ready for a review yet. There are still a few things left to do, so I will post a comment once it looks good to go.

jamesmillerburgess commented:
@hwillson 

I think this is ready for a review!

I've cleaned up the history to be one commit per package and added a few notes, especially where I fixed a few things I found. Hopefully that makes it a bit easier to go through. Again, if there is anything I went overboard with code style-wise then I am happy to go back and rework it.

One thing I noticed is that the test coverage on these packages has some gaps - even with all the TinyTest tests passing, there were a lot of things I caught using my own little fixtures app and then later the `accounts-ui-viewer` app I found in `/examples/unfinished/`. The whole accounts system could benefit from additional unit tests and browser automation tests. I'm happy to work on that, but I don't know if that is something for this PR or not - it would probably take me another week or so, depending on my availability.

hwillson commented:
Awesome @jamesmillerburgess, I'll kick of a review shortly! And thanks for the heads up about the tests. Changes that improve test coverage are definitely welcome. Maybe hold off on submitting changes to expand the test suite until we get through this PR, since there will likely be some back and forth here.

zimme commented:
I think we should try and mention some of the Meteor package maintainers that have accounts related packages, which are widely used, so that they get a heads up on this work and can check that the changes doesn't break their packages.

jamesmillerburgess commented:
@zimme 

Do you have any particular packages in mind? I can run them from this branch and manually test as a first pass.

zimme commented:
@jamesmillerburgess Maybe the `useraccounts:*` and `std:accounts-ui` packages? Some of the community provided `accounts` provides for different services like linkedin etc.

jamesmillerburgess commented:
@zimme Thanks, I'll take a look!

jamesmillerburgess commented:
@zimme I'm having some trouble getting the `useraccounts:*` packages to work with a modern version of meteor, even without my modifications to accounts. It looks like these packages haven't been updated in a few years, and the boilerplate apps I found are using meteor 1.0.3.2. Are you using these packages in a meteor 1.6+ app?

The `std:accounts-ui` package seems to work the same in both 1.6.0.1 and this branch, at least for the things I tried.

@hwillson While investigating these community packages I found a few modernization gaps and some incorrect `const` statements in an `isCordova` section of the packages that isn't exercised by tests or the example app. I've corrected these and squashed them into their respective commits by package.

StorytellerCZ commented on packages/oauth/oauth_server.js:
```diff
@@ -30,58 +30,57 @@ OAuth._requestHandlers = {};
 
-OAuth._loginStyleFromQuery = function (query) {
-  var style;
+OAuth._loginStyleFromQuery = quer => {
```
Typo here. Missing `y`. Should be `query`.

hwillson replied:
Wow, good 👀 @StorytellerCZ!

jamesmillerburgess commented:
I need to take another pass on this one and implement all the feedback I've gotten on the other modernization PRs. I will pick it up again once we get #9656 merged.

StorytellerCZ commented:
@jamesmillerburgess Any update on this?

jamesmillerburgess commented:
Unfortunately, I haven’t done anything else here. When I first worked on
this I had a long period of free time, and now I have very little at all.

This PR probably got way too large to go through the typical review and
merge process. I also added a number of enhancements besides just
modernizing the packages, and I now know that these are best split into
individual PRs.

I’m perfectly happy for anyone to pick up where I left off and get this
merged, whether as a whole or in pieces. Also, if I do happen to get enough
time at some point, I will post an update here.

On Thu, May 3, 2018 at 7:18 PM Jan Dvorak <notifications@github.com> wrote:

> @jamesmillerburgess <https://github.com/jamesmillerburgess> Any update on
> this?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/meteor/meteor/pull/9558#issuecomment-386331714>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/APOnDAr3RZS1N8SuLsa3OVqAkGg872XMks5tux-rgaJpZM4Rl28t>
> .
>

StorytellerCZ commented:
I'll see if I can find some time to get this ready to be merged. Even if not everything that @jamesmillerburgess wanted to do is finished, this is still a great push forward.

StorytellerCZ commented:
So now we have all tests passing. From what I have seen there are differences between devel and the 1.7 release branch.
How should we move forward? Merge into 1.7 branch or wait for it to merge into devel and then aim for a release in 1.7.1?

StorytellerCZ commented:
Fixed the failing test: https://github.com/jamesmillerburgess/meteor/pull/4

StorytellerCZ commented:
Everything is green! Anything else that I can help with? @abernix

jancurn commented:
Hey guys, please can you merge this PR? Since we got rid of the `underscore` Meteor package in our project the GitHub OAuth integration stopped working due to the following error:

```
Uncaught ReferenceError: _ is not defined
    at Object.Github.requestCredential (github-oauth.js?hash=16c1665f9879dcbdba7a80931d0b861525e41b1b:56)
    at AccountsClient.loginWithGithub (github.js:12)
    at AccountsClient.applyLoginFunction (accounts_client.js:108)
    at Meteor.loginWithGithub (github.js:16)
    at Object.click button (useraccounts_core.js?hash=3655f04ee93044d46fc8570d79748d730d5911fb:2547)
    at blaze.js?hash=a1ff2d6d5ecd59ee11e2ba260b8650a9d1140f59:3775
    at Function.Template._withTemplateInstanceFunc (blaze.js?hash=a1ff2d6d5ecd59ee11e2ba260b8650a9d1140f59:3744)
    at Blaze.View.<anonymous> (blaze.js?hash=a1ff2d6d5ecd59ee11e2ba260b8650a9d1140f59:3774)
    at blaze.js?hash=a1ff2d6d5ecd59ee11e2ba260b8650a9d1140f59:2617
    at Object.Blaze._withCurrentView (blaze.js?hash=a1ff2d6d5ecd59ee11e2ba260b8650a9d1140f59:2271)
    at Blaze._DOMRange.<anonymous> (blaze.js?hash=a1ff2d6d5ecd59ee11e2ba260b8650a9d1140f59:2616)
    at HTMLButtonElement.<anonymous> (blaze.js?hash=a1ff2d6d5ecd59ee11e2ba260b8650a9d1140f59:863)
    at HTMLDivElement.dispatch (jquery.js?hash=e15504227515793c40a6a3a8ad340399cb657500:5282)
    at HTMLDivElement.elemData.handle (jquery.js?hash=e15504227515793c40a6a3a8ad340399cb657500:4934)
```

I believe this PR fixes this problem.

benjamn commented:
There's quite a queue of CI test runs at the moment, but I will merge this PR when the tests pass.

StorytellerCZ commented:
@jamesmillerburgess That is a risk with every release. Since these packages will go through beta and rc for 1.7.1, first we have a good chance to catch them there. I will switch to the 1.7.1 release branch later this week to give this a more through test on my staging.
